core/amber/src/main/scala/edu/uci/ics/texera/web/resource/UserDictionaryResource.scala

```diff
@@ -0,0 +1,233 @@
+
+/**
+  * This class handles requests to read and write the user dictionary,
+  * an abstract collection of (key, value) string pairs that is unique for each user
+  * This is accomplished using a mysql table called user_dictionary.
+  * The details of user_dictionary can be found in /core/scripts/sql/texera_ddl.sql
+  */
+@Path("/users/dictionary")
+@Consumes(Array(MediaType.APPLICATION_JSON))
+@Produces(Array(MediaType.APPLICATION_JSON))
+class UserDictionaryResource {
+  final private val userDictionaryDao = new UserDictionaryDao(
+    SqlServer.createDSLContext.configuration
+  )
+
+  /**
+    * This method either retrieves all of the current in-session user's dictionary
+    * as a json object or retrieves the value corresponding to the "key" attribute of the GetRequest.
+    * @param session HttpSession
+    * @param req GetRequest (key may be null for GetDictRequest)
+    * @return Response with a UserDictionaryResponse payload, 200 ok - payload: EntryResponse or DictResponse,
+    * 401 unauthorized - payload: ErrorResponse("invalid session"),
+    * 400 bad request - payload: ErrorResponse("Unable to process JSON"),
+    * 422 Unprocessable Entity - payload: ErrorResponse("no such entry") (if no entry exists for provided key)
+    */
+  @GET
+  def getValue(@Session session: HttpSession, req: GetRequest): Response = {
+    val user = UserResource.getUser(session).orNull
+    if (user == null)
+      Response.status(Response.Status.UNAUTHORIZED).entity(ErrorResponse("invalid session")).build()
+    else if (req == null) {
+      Response
+        .status(Response.Status.BAD_REQUEST)
+        .entity(ErrorResponse("Unable to process JSON"))
+        .build()
+    } else if (req.requestType == GetRequest.GET_ENTRY) {
+      if (!dictEntryExists(user, req.key))
+        Response.status(422).entity(ErrorResponse("no such entry")).build()
+      else Response.ok(EntryResponse(getValueByKey(user, req.key))).build()
+    } else {
+      Response.ok(DictResponse(getDict(user))).build()
+    }
+  }
+
+  /**
+    * This method creates or updates an entry in the current in-session user's dictionary based on
+    * the "key" and "value" attributes of the PostRequest
+    * @param session HttpSession
+    * @param req PostRequest
+    * @return Response, 200 ok - payload: ConfirmationResponse("created" or "updated"),
+    * 401 unauthorized - payload: ErrorResponse("invalid session"),
+    * 400 bad request - payload: ErrorResponse("Unable to process JSON"),
+    */
+  @POST
+  def setValue(
+      @Session session: HttpSession,
+      req: PostRequest
+  ): Response = {
+    val user = UserResource.getUser(session).orNull
+    if (user == null)
+      Response.status(Response.Status.UNAUTHORIZED).entity(ErrorResponse("invalid session")).build()
+    else if (req == null) {
+      Response
+        .status(Response.Status.BAD_REQUEST)
+        .entity(ErrorResponse("Unable to process JSON"))
+        .build()
```

Comment:
I tried this out, but decided against using @NotNull because I couldn't figure out how to return a custom JSON payload on error. I'd prefer that all my error responses have the same format.

I believe there's a mechanism in higher versions of Jackson that provides these things using annotations (@ JsonCreator & @ JsonProperty), but otherwise, I couldn't figure out how to implement validation _cleanly_